package/src/activation.ts

```diff
@@ -0,0 +1,462 @@
+            selector,
+            createReferencesProvider(
+                lsifProviders,
+                searchProviders,
+                lspProviders
+            )
+        )
+    )
+    ctx.subscriptions.add(
+        sourcegraph.languages.registerHoverProvider(
+            selector,
+            createHoverProvider(lsifProviders, searchProviders, lspProviders)
+        )
+    )
+
+    if (lspProviders) {
+        const externalReferencesProvider = lspProviders.externalReferences
+        const implementationsProvider = lspProviders.implementations
+
+        if (externalReferencesProvider) {
+            registerExternalReferencesProvider(
+                ctx,
+                selector,
+                externalReferencesProvider
+            )
+        }
+
+        if (implementationsProvider) {
+            registerImplementationsProvider(
+                ctx,
+                selector,
+                implementationsProvider
+            )
+        }
+    }
+}
+
+function createDefinitionProvider(
+    lsifProviders: LSIFProviders,
+    searchProviders: SearchProviders,
+    lspProviders?: LSPProviders
+): sourcegraph.DefinitionProvider {
+    async function* provideDefinition(
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ): AsyncGenerator<
+        sourcegraph.Definition | null | undefined,
+        void,
+        undefined
+    > {
+        const lsifResult = await lsifProviders.definition(doc, pos)
+        if (lsifResult) {
+            yield lsifResult.value
+            return
+        }
+
+        if (lspProviders) {
+            yield* lspProviders.definition(doc, pos)
+            return
+        }
+
+        let searchResult = await searchProviders.definition(doc, pos)
+        if (!searchResult) {
+            yield undefined
+            return
+        }
+
+        if (!Array.isArray(searchResult)) {
+            yield { ...searchResult, badge: impreciseBadge }
+            return
+        }
+
+        yield searchResult.map(v => ({ ...v, badge: impreciseBadge }))
+    }
+
+    return {
+        provideDefinition: memoizePrevious(areProviderParamsEqual, (doc, pos) =>
+            observableFromAsyncGenerator(
+                abortPrevious(() => provideDefinition(doc, pos))
+            ).pipe(shareReplay(1))
+        ),
+    }
+}
+
+function createReferencesProvider(
+    lsifProviders: LSIFProviders,
+    searchProviders: SearchProviders,
+    lspProviders?: LSPProviders
+): sourcegraph.ReferenceProvider {
+    // Gets an opaque value that is the same for all locations
+    // within a file but different from other files.
+    const file = (loc: sourcegraph.Location) =>
+        `${loc.uri.host} ${loc.uri.pathname} ${loc.uri.hash}`
+
+    async function* provideReferences(
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position,
+        ctx: sourcegraph.ReferenceContext
+    ): AsyncGenerator<sourcegraph.Location[] | null, void, undefined> {
```

Comment:
Note: this logic isn't great but it's what currently exists for Go and Typescript. We should favor fast LSIF results first, then LSP results, then search results (if LSP is provided). This is possible now that we've updated our providers to be AsyncGenerators